### ISSUE TITLE
@jyegerlehner (or anyone), how long did it take to train all the audio data??

### ISSUE THREAD

**Issue by aclaussen1** [ibab/tensorflow-wavenet#193]
Other questions:
1) What hardware are you using?
2) Are you using GPU for tensorflow

**Comment by dunnevan** [comment]
1.  The size of the network (wavenet-params) is going to affect the speed.  On my 1070 GPU a step is often in the ballpark of 0.8 seconds.

2.  One step is one sample of audio.  If you have 16000 samples a second then 20 steps is next to nothing.  With that said, the  loss is more important to look at.  This 'loss' is the training loss, we don't compute a validation loss.  After 64000 steps (4 seconds of audio), the network I'm training was at ~1.2 loss.  Of course it is going to depend on your training data and parameters.

**Comment by jyegerlehner** [comment]
@aclaussen1 

1. Titan XP. I build tensorflow from source, and haven't had an unsolvable problem running on GPU. But setup and install can be a pain. I've gotten used to working through it.

2. As dunnevan pointed out, you're not getting anywhere with 20 steps/ hour, unless you have a year to wait.  5.438 loss is about the initial un-trained loss that I see. I've gotten "OK" results, for sufficiently small values of OK, at about 30K steps training on a single speaker, and loss would be somewhere around 2.0, and better results as loss gets down to around 1.0->1.5, as dunnevan suggests. I've gotten my best results after a week or so at around 500K steps. Which, by the way, is vaguely similar to what the Montreal guys reported in [sample RNN](https://openreview.net/forum?id=SkxKPDv5xl) which appears to be closely related to WaveNet.

And it can all depend on the many knobs available to turn. Silence trimming seems to be important, with better results turning it off or way down. And other things that have been discussed in the various PR and issue threads around here I don't have time to go dig up.

**Comment by tszumowski** [comment]
1. When running on the CPU of a NVIDIA DIGITS box, it takes about 60 seconds per step. When I use the Titan X GPU on the box, it takes about 1 second per step.

2. Right now I'm trying to replicate the results of an ipython notebook in the [fast-wavenet](https://github.com/tomlepaine/fast-wavenet) (of which the generator from there was ported over to here). Running the fast-wavenet code, it completes training in about 5 minutes on the GPU. The test file there under assets/voice.wav is encoded as 44.1K. I re-encoded it to 16K to keep the same default parameters here. That wav file is only 1 second long so it trains quick. I get down to loss=0.05 within 100 steps. However, when I generate I don't get a good sounding waveform (still debugging that).

@dunnevan I just wanted to make sure I understand the steps properly. You stated above: 
> One step is one sample of audio. If you have 16000 samples a second then 20 steps is next to nothing.

My understanding from train.py and audio_reader.py's thread_main(), is that a sample in terms of training is defined by args.sample_size, which defaults to 100000. So a **sample in terms of training** is actually 100000 **audio samples**?

**Comment by veqtor** [comment]
I also thought that a step encompassed trying to fit parameters to args.sample_size values, because when using convnets as wavenet does, it can adjust to a chunk of output values per training step. Isn't this correct? I thought this was a strength of the wavenet architecture.

**Comment by nakosung** [comment]
I have tested smaller sample size with minibatch, it turned out similar.

2016년 12월 14일 (수) 오후 6:20, Göran Sandström <notifications@github.com>님이 작성:

**Comment by veqtor** [comment]
Why would we train on 10 gigabytes of audio at 16khz if one step = one sample?
Also, if one step = one sample, then at 400k you would have trained 25 seconds of audio, which couldn't be nearly enough to differentiate between different speakers.

For google to reach their results on piano, which is 10 hours of material, that would mean:
576 BILLION steps! The only explanation I have is a multi-tower network with A LOT of GPU's running in parallel or even TPU's

**Comment by liangmin0020** [comment]
Is the original set "SAMPLE_SIZE = 100000" in the train.py mean only train 100000/16000~6.25s of audio? wave?

**Comment by veqtor** [comment]
I'm trying to figure this out, in the audio reader, if every training step would mean only one sample, then we essentially end up only training on every SAMPLE_SIZE samples since we feed a whole block per training step to the loss function, or am I missing something?

**Comment by jyegerlehner** [comment]
>  if every training step would mean only one sample, then we essentially end up only training on every SAMPLE_SIZE samples since we feed a whole block per training step to the loss function

Not just one sample. On each training step, every one of the many thousands of samples in the input is also a target for the output. See [this line](https://github.com/ibab/tensorflow-wavenet/blob/master/wavenet/model.py#L633), and the code and comment above it: the target for the cross-entropy loss is just the entire input tensor shifted forward one time step.

**Comment by ianni67** [comment]
@tszumowski 

>  However, when I generate I don't get a good sounding waveform (still debugging that).

I'm also testing the fast-wavenet pynb and get quite strange results. My generated audio is clean in the first fraction of second and then becomes noisy. Did you eventually solve your problems?

**Comment by BuchyLX** [comment]
@jyegerlehner 
Greeting,
As you mentioned above, the network can only train 30sec audio data within one week. So I wonder HOW Google could train through their 30 hours training corpus? Do you any idea the method they used? Train in parallel then merge the model?

**Comment by DabiaoMa** [comment]
The number of wav file processed during one training step is only determined by the parameter BATCH_SIZE. This model splits one wav file into many sub parts and enqueue them all to the tensorflow queue. While sample_size is not necessarily larger than receptive field size because the size of the small parts is (sample_size + receptive field size).

**Comment by DiyuanLu** [comment]
> Is the original set "SAMPLE_SIZE = 100000" in the train.py mean only train 100000/16000~6.25s of audio? wave?
to my understanding, they split input audio file into chunks of size roughly 6.25s, feed it to the network, and get the loss, and then use it in the optimizer which tries to minimize the loss.
Here you can find a clear explanation http://sergeiturukin.com/2017/03/02/wavenet.html

**Comment by DiyuanLu** [comment]
@dunnevan 

> One step is one sample of audio. If you have 16000 samples a second then 20 steps is next to nothing. With that said, the loss is more important to look at. This 'loss' is the training loss, we don't compute a validation loss. After 64000 steps (4 seconds of audio), the network I'm training was at ~1.2 loss. Of course it is going to depend on your training data and parameters.

64000steps is 4 seconds of audio, or 64000 * SAMPLE_SIZE(1000000) / 16000 seconds?
I am a bit confused here.
